File: bin/env_manager.py
#!/usr/bin/env python
# -*- coding=utf8 -*-
import re
import os
import json
import subprocess
from dotenv.main import DotEnv


class EnvManage(object):

    def __init__(self, env_path):
        if os.path.isfile(env_path) is False:
            raise ValueError("env path '%s' is not a file" % env_path)
        self.env_path = env_path
        self.envs = self.read_envs()
        blank_line_regex = r"^\s*$"
        annotation_line_regex = r"^#\s[^\[]*$"
        self.blank_line_regexp = re.compile(blank_line_regex)
        self.annotation_line_regexp = re.compile(annotation_line_regex)

    def read_envs(self):
        manager = DotEnv(self.env_path)
        envs = manager.dict()
        return envs

    def show_sections(self):
        sections = set()
        for env in self.envs:
            section = env.split("_")[0]
            sections.add(section)
        sections = list(sections)
        print(json.dumps(sections, ensure_ascii=False, indent=4))
        return sections

    def show_envs(self, indent=4):
        data = json.dumps(self.envs, ensure_ascii=False, indent=indent)
        print(data)
        return self.envs

    def search_envs(self, keyword, indent=4):
        """
        检索以关键词开头的env
        :param keyword: 关键词
        :param indent: json序列化时的缩进参数
        :return: 包含关键词的配置项
        """
        result = dict()
        for key, val in self.envs.items():
            if key.startswith(keyword):
                result[key] = val
        data = json.dumps(result, ensure_ascii=False, indent=indent)
        print(data)
        return result

    def search_env(self, name):
        value = self.envs.get(name)
        if value is None:
            print("env文件中未包含配置项：%s" % name)
        else:
            print("%s的当前值为: '%s'" % (name, value))
        return value

    def add_section(self, section, check=True):
        """
        追加板块
        :param section: 板块
        :param check: 追加验证是否已存在
        :return: bool
        """
        if check is True:
            print("start check section %s is exist" % section)
            line_no = self.has_section(section)
            if line_no:
                print("add section failed: section %s already exist" % section)
                return False
        end_line = subprocess.getoutput('tail -1 %s' % self.env_path)
        print("start add section %s" % section)
        # 如果文件结尾非空行,则先追加一个空行
        if not self.blank_line_regexp.match(end_line):
            cmd = 'echo "" >> %s' % self.env_path
            subprocess.call(cmd, shell=True)
        cmd = 'echo "# [%s]" >> %s' % (section, self.env_path)
        subprocess.call(cmd, shell=True)
        line_no = self.has_section(section)
        if line_no is False:
            print("add section %s failed" % section)
            return False
        else:
            print("add section %s succeed" % section)
            return True

    def has_section(self, section):
        """
        检查是否存在指定板块
        :param section: 板块
        :return: bool
        """
        cmd = r'sed -n "/^# \[%s\]/=" %s' % (section, self.env_path)
        line_no = subprocess.getoutput(cmd)
        if line_no == "":
            print("section %s not found" % section)
            return False
        print("section %s on the line: %s" % (section, line_no))
        return line_no

    def del_section(self, section):
        """
        删除板块及板块下的所有env
        :param section: 板块名
        :return:
        """
        exist = self.has_section(section)
        if exist is False:
            print("del section failed: section %s not found" % section)
            return False
        # section所在的行
        lines = [int(line) for line in exist.split()]
        # 删除section上面的空行
        for line in lines:
            # 是在第一行则跳过
            if line == 1:
                continue
            pre_line = line - 1
            # 查看上一行的内容
            pre_line_content_cmd = "sed -n '%dp' %s" % (pre_line, self.env_path)
            pre_line_content = subprocess.getoutput(pre_line_content_cmd)
            # 如果上一行不为空则跳过
            if not self.blank_line_regexp.match(pre_line_content):
                continue
            # 删除上面的空行
            del_pre_line_cmd = "sed -i '%dd' %s" % (pre_line, self.env_path)
            subprocess.call(del_pre_line_cmd, shell=True)
        cmd = r'sed -i "/^# \[%s\]/d" %s' % (section, self.env_path)
        subprocess.call(cmd, shell=True)
        exist = self.has_section(section)
        if exist is False:
            print("del section %s succeed" % section)
            print("start del envs about the section %s" % section)
            envs = self.search_envs(section)
            for env in envs:
                self.del_env(env)
            return True
        else:
            print("del section %s failed" % section)
            return False

    def del_env(self, name):
        """
        删除一个env
        :param name: 要删除的env名
        :return:
        """
        exist = self.search_env(name)
        if exist is None:
            print("del env failed: env %s not found" % name)
            return False
        line_cmd = 'sed -n "/^%s=/=" %s' % (name, self.env_path)
        line = int(subprocess.getoutput(line_cmd))
        pre_line = line - 1
        # 查看上一行的内容
        pre_line_content_cmd = "sed -n '%dp' %s" % (pre_line, self.env_path)
        pre_line_content = subprocess.getoutput(pre_line_content_cmd)
        # 如果上面一行时注释, 则删除注释
        if self.annotation_line_regexp.match(pre_line_content):
            del_pre_line_cmd = "sed -i '%dd' %s" % (pre_line, self.env_path)
            subprocess.call(del_pre_line_cmd, shell=True)
        # 删除配置行
        cmd = 'sed -i "/^%s=/d" %s' % (name, self.env_path)
        subprocess.call(cmd, shell=True)
        # 判断是否删除成功
        new_envs = self.read_envs()
        value = new_envs.get(name)
        if value is None:
            print("del env %s succeed" % name)
            return True
        else:
            print("del env %s failed, the value is '%s'" % (name, value))
            return False

    def add_env(self, section, name, value):
        """
        追加一个env
        :param section: 指定追加到哪个部分下面
        :param name: 要追加的变量名
        :param value: 变量的值
        :return: 是否更改成功 -> bool
        """
        if name == "" or section == "":
            print("add env failed: param name and section can't is ''")
            return False
        name = "_".join([section, name])
        # 判断env是否已存在
        exist = self.search_env(name)
        # 如果已存在则进行更新
        if exist is not None:
            print("env %s is already exist, will be updated" % name)
            return self.update_env(name, value)
        line_no = self.has_section(section)
        # 如果板块不存在则追加板块
        if line_no is False:
            self.add_section(section, check=False)
        # 追加env到板块下的命令
        cmd = r'sed -i "/^# \[%s\]/a%s=%s" %s' % (section, name, value, self.env_path)
        # 执行追加命令
        subprocess.call(cmd, shell=True)
        # 判断是否追加成功
        new_envs = self.read_envs()
        value = new_envs.get(name)
        if value is not None:
            print("add env %s succeed, value is '%s'" % (name, value))
            return True
        else:
            print("add env %s failed" % name)
            return False

    def update_env(self, name, value):
        """
        更新配置
        :param name: 要更新的配置项
        :param value: 新值
        :return: 是否更改成功 -> bool
        """
        ret = False
        exist = self.search_env(name)
        if exist is None:
            print("修改失败, 该方法不支持新增配置项")
        elif exist == value:
            print("不需要修改,配置项%s要修改的值%s跟原来的值%s是一致的" % (name, value, exist))
        else:
            print("开始修改配置")
            cmd = 'sed -i "/^%s=/c%s=%s" %s' % (name, name, value, self.env_path)
            output = subprocess.getoutput(cmd)
            new_envs = self.read_envs()
            value = new_envs.get(name)
            if value is not None and output == '':
                print("修改配置%s成功, 修改后的值为: '%s'" % (name, value))
                # 如果是修改的进程数配置,则设置进程数
                if name.startswith("circus_") and name.endswith("_num"):
                    server = name.replace("circus_", "").replace("_num", "")
                    cmd = 'circusctl set %s numprocesses %s' % (server, value)
                    set_ret = subprocess.getoutput(cmd)
                    if "error" not in set_ret:
                        print('设置服务%s的进程数为%s成功' % (server, value))
                    else:
                        print('设置服务%s的进程数为%s失败: %s' % (server, value, set_ret))
                ret = True
            else:
                print("修改配置%s失败: %s" % (name, output))
        return ret


if __name__ == '__main__':
    from argparse import ArgumentParser
    from importlib import import_module

    import_module("__init__")
    from common.conts import ENV_PATH

    desc = "envs manager"
    parser = ArgumentParser(description=desc)
    parser.add_argument("-ls", "--list_sections", help="show sections", action='store_true')
    parser.add_argument("-l", "--list", help="show envs", action='store_true')
    parser.add_argument("-sf", "-fs", "--fuzzy_search", help="模糊查询env的值, 将查询以指定关键词开头的env")
    parser.add_argument("-f", "--fuzzy", help="是否是模糊查询,只能配合-s/--search使用", action='store_true')
    parser.add_argument("-s", "--search", type=str, help="查询env的值")
    parser.add_argument("-c", "--clear", type=str, help="删除指定板块下的所有env")
    parser.add_argument("-u", "--update", type=str, help="更新env的值,接收两个参数name、value", nargs=2)
    parser.add_argument("-d", "--delete", type=str, help="删除一个env")
    parser.add_argument("-a", "--add", type=str,
                        help="追加一个env,接收三个参数section、name、value,将追加一个section_name=value的env", nargs=3)
    parser.add_argument("-p", "--path", type=str, default=ENV_PATH, help="env文件的路径")
    args = parser.parse_args()
    env_path = args.path
    env_manager = EnvManage(env_path)
    list_envs = args.list
    list_sections = args.list_sections
    fuzzy_keyword = args.fuzzy_search
    keyword = args.search
    update = args.update
    add = args.add
    delete = args.delete
    clear = args.clear

    if list_envs is True:
        env_manager.show_envs()
    elif list_sections:
        env_manager.show_sections()
    elif fuzzy_keyword or keyword:
        keyword = fuzzy_keyword or keyword
        fuzzy = fuzzy_keyword and True or args.fuzzy
        if fuzzy:
            env_manager.search_envs(keyword)
        else:
            env_manager.search_env(keyword)
    elif update:
        env_manager.update_env(*update)
    elif add:
        env_manager.add_env(*add)
    elif delete:
        env_manager.del_env(delete)
    elif clear:
        env_manager.del_section(clear)
    else:
        print("invalid cmd, please read the help doc")
        parser.print_help()

File: deploy/env_manager.go
/*
 * @Author: xiangcai
 * @Date: 2022-02-28 10:00:08
 * @LastEditors: xiangcai
 * @LastEditTime: 2022-03-02 16:30:33
 * @Description: env manager
 * 	macos需安装配置gnu-sed兼容sed命令
 */

package main

import (
	"bytes"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"

	"github.com/sgs921107/gcommon"
)

var (
	circusPrefix = "circus_"
	circusSuffix = "_num"
)

func printf(format string, a ...interface{}) {
	println(fmt.Sprintf(format, a...))
}

// 执行shell
func execShell(command string) ([]byte, error) {
	return exec.Command("/bin/bash", "-c", command).Output()
}

// 复制文件
func copyFile(srcFilePath, dstFilePath string) (int64, error) {
	// 打开源文件
	srcFile, err := os.Open(srcFilePath)
	if err != nil {
		return 0, err
	}
	defer srcFile.Close()
	dstFile, err := os.OpenFile(dstFilePath, os.O_RDWR|os.O_CREATE, 0644)
	if err != nil {
		return 0, err
	}
	defer dstFile.Close()
	srcMD5, err := fileMD5(srcFile)
	if err != nil {
		return 0, err
	}
	dstMD5, err := fileMD5(dstFile)
	if err != nil {
		return 0, err
	}
	// 如果文件内容一样则不执行复制
	if srcMD5 == dstMD5 {
		return 0, nil
	}
	// 游标移动至原点
	if _, err := srcFile.Seek(0, 0); err != nil {
		return 0, err
	}
	if _, err := dstFile.Seek(0, 0); err != nil {
		return 0, err
	}
	// 复制文件内容
	if written, err := io.Copy(dstFile, srcFile); err != nil {
		return 0, err
	} else {
		// 截断
		return written, dstFile.Truncate(written)
	}
}

// 文件MD5
func fileMD5(fp *os.File) (string, error) {
	cipherBytes := [md5.Size]byte{}
	for {
		buf := make([]byte, 1024)
		n, err := fp.Read(buf)
		if n > 0 {
			cipherBytes = md5.Sum(append(cipherBytes[:md5.Size], buf[:n]...))
		}
		if err == nil {
			continue
		} else if err == io.EOF {
			break
		} else {
			return "", err
		}
	}
	return hex.EncodeToString(cipherBytes[:md5.Size]), nil
}

type EnvManager struct {
	// env配置文件的路径
	envPath string
	// 对临时文件进行操作成功后回写至env path
	tmpPath string
	// 空行
	blankLineRegexp  *regexp.Regexp
	annotationRegexp *regexp.Regexp
}

// 加载env
func (m EnvManager) envs() map[string]string {
	envs, err := gcommon.ReadFileEnv(m.tmpPath)
	if err != nil {
		panic(fmt.Sprintf("Load Envs Error: %s", err.Error()))
	}
	return envs
}

// 以json格式输出
func (m EnvManager) jsonPrint(data interface{}) error {
	b, err := json.Marshal(data)
	if err != nil {
		return err
	}
	var output bytes.Buffer
	if err := json.Indent(&output, b, " ", "\t"); err != nil {
		return err
	}
	fmt.Println(output.String())
	return nil
}

func (m EnvManager) showSections() {
	sectionsMap := make(map[string]int)
	for key := range m.envs() {
		section := strings.Split(key, "_")[0]
		sectionsMap[section]++
	}
	sections := make([]string, 0)
	for section := range sectionsMap {
		sections = append(sections, section)
	}
	if err := m.jsonPrint(sections); err != nil {
		printf("Show Sections Error: %s", err.Error())
	}
}

// 展示所有的env
func (m EnvManager) showEnvs() {
	if err := m.jsonPrint(m.envs()); err != nil {
		printf("Show Envs Error: %s", err.Error())
	}
}

// 查询包含指定内容的env
func (m EnvManager) sectionEnvs(section string) map[string]string {
	envs := make(map[string]string)
	for key, val := range m.envs() {
		if strings.HasPrefix(key, section) {
			envs[key] = val
		}
	}
	m.jsonPrint(envs)
	return envs
}

// 获取env的值
func (m EnvManager) get(name string) (string, bool) {
	val, ok := m.envs()[name]
	if ok {
		printf("env %s value: '%s'", name, val)
	} else {
		printf("env %s not found", name)
	}
	return val, ok
}

// 更新env
func (m EnvManager) update(name, value string) bool {
	if oriVal, ok := m.get(name); !ok {
		printf("update env %s failed: env not found", name)
		return false
	} else if oriVal == value {
		printf("env %s not need update, the value %s same as ori value %s", name, value, oriVal)
		return false
	}
	cmd := fmt.Sprintf("sed -i '/^%s=/c%s=%s' %s", name, name, value, m.tmpPath)
	if output, err := execShell(cmd); err != nil {
		printf("update env %s error: %s", name, err.Error())
		return false
	} else {
		printf("begin update env %s", name)
		if curVal, _ := m.get(name); curVal != value {
			printf("update env %s failed: %s", name, string(output))
			return false
		} else {
			if _, err := copyFile(m.tmpPath, m.envPath); err != nil {
				printf("update env %s failed, overwrite failed: %s", name, err.Error())
				return false
			}
			printf("update env %s succeed, cur value: '%s'", name, curVal)
			// 如果是circus管理的watcher进程数配置, 则设置进程数
			if strings.HasPrefix(name, circusPrefix) && strings.HasSuffix(name, circusSuffix) {
				watcher := strings.TrimSuffix(strings.TrimPrefix(name, circusPrefix), circusSuffix)
				circusCmd := fmt.Sprintf("circusctl set %s numprocesses %s", watcher, curVal)
				if output, err := execShell(circusCmd); err != nil {
					printf("set numprocesses of circus watcher %s to %s error: %s", watcher, curVal, err.Error())
				} else if strings.Contains(string(output), "error") {
					printf("set numprocesses of circus watcher %s to %s failed: %s", watcher, curVal, string(output))
				} else {
					printf("set numprocesses of circus watcher %s to %s succeed", watcher, curVal)
				}
			}
			return true
		}
	}
}

// 添加env
func (m EnvManager) addEnv(section, name, value string) bool {
	if section == "" || name == "" {
		printf("add env failed, section[%s] and name[%s] must not be empty string", section, name)
		return false
	}
	name = section + "_" + name
	if _, ok := m.get(name); ok {
		printf("env %s already exist, will be updated", name)
		return m.update(name, value)
	}
	// 尝试新增板块
	m.addSection(section)
	printf("begin add env %s", name)
	// 新增env
	addEnvCmd := fmt.Sprintf(`sed -i '/^# \[%s\]/a%s=%s' %s`, section, name, value, m.tmpPath)
	if output, err := execShell(addEnvCmd); err != nil {
		printf("add env %s error: %s", name, err.Error())
		return false
	} else {
		if curVal, _ := m.get(name); curVal != value {
			printf("add env %s failed: %s", name, string(output))
			return false
		} else {
			if _, err := copyFile(m.tmpPath, m.envPath); err != nil {
				printf("add env %s failed, overwrite failed: %s", name, err.Error())
				return false
			}
			printf("add env %s succeed, value: '%s'", name, curVal)
			return true
		}
	}
}

// 删除env
func (m EnvManager) delEnv(name string) bool {
	if _, ok := m.get(name); !ok {
		printf("del env %s failed: env not found", name)
		return false
	}
	lineCmd := fmt.Sprintf("sed -n '/^%s=/=' %s", name, m.tmpPath)
	linesOutput, err := execShell(lineCmd)
	if err != nil {
		printf("del env %s error: %s", name, err.Error())
		return false
	}
	lines := regexp.MustCompile(`\d+`).FindAll(linesOutput, -1)
	for index := len(lines) - 1; index >= 0; index-- {
		lineNo, err := strconv.Atoi(strings.TrimSpace(string(lines[index])))
		if err != nil {
			printf("del env %s failed: get env lineNo error: %s", name, err.Error())
			return false
		}
		preLine := lineNo - 1
		// 查看上一行的内容
		preLineCmd := fmt.Sprintf("sed -n '%dp' %s", preLine, m.tmpPath)
		// 如果上一行内容是注释行 则删除
		if output, _ := execShell(preLineCmd); m.annotationRegexp.Match(output) {
			delPreLineCmd := fmt.Sprintf("sed -i '%dd' %s", preLine, m.tmpPath)
			execShell(delPreLineCmd)
		}
	}
	// 删除配置行
	delEnvCmd := fmt.Sprintf("sed -i '/^%s=/d' %s", name, m.tmpPath)
	if _, err := execShell(delEnvCmd); err != nil {
		printf("Exec Del Env Cmd Error: %s", err.Error())
		return false
	}
	// 校验是否删除成功
	if curVal, ok := m.get(name); ok {
		printf("del env %s failed, value: '%s'", name, curVal)
		return false
	}
	if _, err := copyFile(m.tmpPath, m.envPath); err != nil {
		printf("del env %s failed, overwrite failed: %s", name, err.Error())
		return false
	}
	printf("del env %s succeed", name)
	return true
}

// 查询板块是否存在
func (m EnvManager) sectionLines(section string) ([]string, error) {
	cmd := fmt.Sprintf(`sed -n '/^# \[%s\]/=' %s`, section, m.tmpPath)
	if lines, err := execShell(cmd); err != nil {
		printf("Exec CMD %s Error: %s", cmd, err.Error())
		return nil, err
	} else {
		linesNo := regexp.MustCompile(`\d+`).FindAllString(string(lines), -1)
		if len(linesNo) == 0 {
			printf("section %s not found", section)
			return nil, nil
		} else {
			printf("section %s line: %s", section, strings.Join(linesNo, ", "))
			return linesNo, nil
		}
	}
}

// 添加板块
func (m EnvManager) addSection(section string) bool {
	if lines, err := m.sectionLines(section); err != nil {
		printf("Get LineNo Of The Section %s Error: %s", section, err.Error())
		return false
	} else if len(lines) > 0 {
		return false
	}
	printf("start add section %s", section)
	endLineCmd := fmt.Sprintf("tail -1 %s", m.tmpPath)
	if endLine, err := execShell(endLineCmd); err != nil {
		printf("Add Section %s Error: get end line content error: %s", section, err.Error())
		return false
		// 如果最后一行不是空行, 添加空行
	} else if !m.blankLineRegexp.Match(endLine) {
		execShell(fmt.Sprintf(`echo "" >> %s`, m.tmpPath))
	}
	// 添加section行
	execShell(fmt.Sprintf(`echo "# [%s]" >> %s`, section, m.tmpPath))
	if lines, _ := m.sectionLines(section); len(lines) > 0 {
		printf("add section %s succeed", section)
		return true
	} else {
		printf("add section %s failed", section)
		return false
	}
}

// 删除板块
func (m EnvManager) delSection(section string) bool {
	if lines, _ := m.sectionLines(section); len(lines) == 0 {
		printf("del section %s failed: get lineno failed", section)
		return false
		// 删除上面的空行
	} else {
		for index := len(lines) - 1; index >= 0; index-- {
			lineNo, err := strconv.Atoi(strings.TrimSpace(string(lines[index])))
			if err != nil {
				printf("del section %s failed: get section lineNo error: %s", section, err.Error())
				return false
			}
			// 如果是在第一行则跳过
			if lineNo == 1 {
				continue
			}
			preLine := lineNo - 1
			// 查看上一行的内容
			preLineCmd := fmt.Sprintf("sed -n '%dp' %s", preLine, m.tmpPath)
			// 如果上一行内容是空行 则删除
			if output, _ := execShell(preLineCmd); m.blankLineRegexp.Match(output) {
				delPreLineCmd := fmt.Sprintf("sed -i '%dd' %s", preLine, m.tmpPath)
				execShell(delPreLineCmd)
			}
		}
		// 删除板块
		delSectionCmd := fmt.Sprintf(`sed -i "/^# \[%s\]/d" %s`, section, m.tmpPath)
		if output, err := execShell(delSectionCmd); err != nil {
			printf("del section %s error: %s", section, err.Error())
			return false
		} else {
			// 校验是否删除成功
			if lines, _ := m.sectionLines(section); len(lines) > 0 {
				printf("del section %s failed: %s", section, string(output))
				return false
			}
			if _, err := copyFile(m.tmpPath, m.envPath); err != nil {
				printf("del section %s failed, overwrite failed: %s", section, err.Error())
				return false
			}
			printf("del section %s succeed", section)
			printf("start del envs about the section %s", section)
			for env := range m.sectionEnvs(section) {
				m.delEnv(env)
			}
			return true
		}
	}
}

// 生成EnvManager对象
func newEnvManager(envPath, tmpPath string) EnvManager {
	// 复制env到临时文件
	if _, err := copyFile(envPath, tmpPath); err != nil {
		panic(fmt.Sprintf("copy env path to tmpPath error: %s", err.Error()))
	}
	manager := EnvManager{
		envPath: envPath,
		tmpPath: tmpPath,
		// 空白行
		blankLineRegexp: regexp.MustCompile(`^\s*$`),
		// 注释行
		annotationRegexp: regexp.MustCompile(`^\s*#[^\[]*$`),
	}
	return manager
}

func main() {
	var defaultEnvPath, tmpPath = ".env", "/tmp/temp_env"
	var envs, sections, update, add bool
	var env, section, clear, delete, path string
	flag.BoolVar(&envs, "l", false, "-l: list envs")
	flag.BoolVar(&sections, "ls", false, "-ls: list sections")
	flag.StringVar(&env, "s", "", "-s env: search env")
	flag.StringVar(&section, "ss", "", "-ss section: search section envs")
	flag.BoolVar(&update, "u", false, "-u env value: update env")
	flag.BoolVar(&add, "a", false, "-a section name value: add env")
	flag.StringVar(&delete, "d", "", "-d env: del env")
	flag.StringVar(&clear, "c", "", "-c section: clear section envs")
	flag.StringVar(&path, "path", defaultEnvPath, "-path path: env path")
	flag.Parse()
	manager := newEnvManager(path, tmpPath)
	args := flag.Args()
	nArg := flag.NArg()
	if envs {
		manager.showEnvs()
	} else if sections {
		manager.showSections()
	} else if env != "" {
		manager.get(env)
	} else if section != "" {
		manager.sectionEnvs(section)
	} else if add && nArg == 3 {
		manager.addEnv(args[0], args[1], args[2])
	} else if update && nArg == 2 {
		manager.update(args[0], args[1])
	} else if delete != "" {
		manager.delEnv(delete)
	} else if clear != "" {
		manager.delSection(clear)
	} else {
		flag.Usage()
	}
}
